feat: stabilize credential-process and registry-auth

arlosi opened the pull request:
Stabilization PR for `registry-auth` and `credential-process`.

Tracking approval of this stabilization is done in the via the FCP in [#8933](https://github.com/rust-lang/cargo/issues/8933#issuecomment-1711990123). This PR is here to help reviewers of the FCP.

* Stabilizes `registry-auth` and `credential-process`
* Makes authenticated registries require a credential provider
* Adds stable documentation for credential providers and authenticated registries

Closes #8933
Closes #10474

bors commented:
:umbrella: The latest upstream changes (presumably #12648) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

ehuss commented on src/doc/src/reference/registry-authentication.md:
```diff
@@ -0,0 +1,103 @@
+[`cargo logout`]: ../commands/cargo-logout.md
+
+## External providers
+For external providers that follow Cargo's [credential provider protocol](credential-provider-protocol.md)
+use the path to the executable (or the executable name if on the `PATH`).
```
This sentence was a little confusing on my first read. Does something like the following seem a little clearer?

```suggestion
For external providers that follow Cargo's [credential provider protocol](credential-provider-protocol.md),
the configuration value should be a string with the path to the executable (or the executable name if on the `PATH`).
```

arlosi replied:
Yes, this is better. Thanks!

ehuss commented:
Thanks so much for writing all the documentation, it looks great!

Just some minor notes about the documentation:

Can you update https://github.com/rust-lang/cargo/blob/HEAD/src/doc/src/reference/environment-variables.md to include `CARGO_CREDENTIAL_ALIAS_<name>` and `CARGO_REGISTRIES_<name>_CREDENTIAL_PROVIDER` and `CARGO_REGISTRY_GLOBAL_CREDENTIAL_PROVIDERS`?

----

One thing I'd like to chat about is documenting the stability of the built-in credential providers. I'm concerned that we may not be able to support them indefinitely, at which point we would need to remove one which would break cargo's general compatibility guarantees. For example, let's say Apple or Microsoft suddenly start requiring signed binaries in order to access the global credential stores, and we can't accommodate that. Or, let's say libsecret stops development and distros stop using it. I'm wondering it there should be a note in the auth or config chapter about that? I don't want a note that sounds scary, since it is unlikely that we would stop supporting them, and we would also try to provide a warning ahead of time if possible if we plan to remove it. But I do think it is useful to clearly communicate to users what our expectations around compatibility are.

An alternate idea would be to have a separate "compatibility contract" that would generally state how cargo treats backwards and forwards compatibility, and then exceptions like this (and other things like RUSTFLAGS or platform support) could be mentioned there. Right now, we only have https://doc.crates.io/contrib/design.html which spells that out, but that is not end-user documentation. Perhaps that could move to the cargo guide?

Any thoughts on that?

arlosi commented:
Good catch on the missing docs in `environment-variables.md`. Updated.

> I'm concerned that we may not be able to support them indefinitely, at which point we would need to remove one which would break cargo's general compatibility guarantees.

I've added a note in `registry-authentication.md` that says:
> Cargo includes several built-in credential providers. The available built-in providers may change in future Cargo releases.

Do you think that's sufficient?

ehuss commented:
> Do you think that's sufficient?

Seems good to me.

Eh2406 commented:
> Cargo includes several built-in credential providers. The available built-in providers may change in future Cargo releases. Although, we do not have plans to do so in the near future. If plans change we will update this documentation.

Or sum such.

ehuss commented:
I forgot, can you also update [`publish.py`](https://github.com/rust-lang/cargo/blob/master/publish.py) so that it includes all of the credential crates? 1password doesn't *need* to be included, but I don't think it will cause any harm to include it.

weihanglo commented:
Your efforts on this are much appreciated, Arlo! Also thanks everyone for reviewing and testing on this feature :)

We're not going to wait for the FCP ended, as the beta branching-off window is approaching, and the FCP [was started 9 days ago](https://github.com/rust-lang/cargo/issues/8933#issuecomment-1711990145). If you encounter any issue with this stabilization, please report it here or open a new issue.

@bors r+

bors commented:
:pushpin: Commit d345ca212f44f6ccd998ac3275529481b6f437f4 has been approved by `weihanglo`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/cargo) for this repository.

<!-- @bors r=weihanglo d345ca212f44f6ccd998ac3275529481b6f437f4 -->
<!-- homu: {"type":"Approved","sha":"d345ca212f44f6ccd998ac3275529481b6f437f4","approver":"weihanglo","queue":"https://bors.rust-lang.org/homu/queue/cargo"} -->

bors commented:
:hourglass: Testing commit d345ca212f44f6ccd998ac3275529481b6f437f4 with merge 7149418ed375d2e73256c7bee675c8ce380794ba...
<!-- homu: {"type":"BuildStarted","head_sha":"d345ca212f44f6ccd998ac3275529481b6f437f4","merge_sha":"7149418ed375d2e73256c7bee675c8ce380794ba"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang/cargo/actions/runs/6217073139/job/16872347864)
Approved by: weihanglo
Pushing 7149418ed375d2e73256c7bee675c8ce380794ba to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"weihanglo","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang/cargo/actions/runs/6217073139/job/16872347864"},"merge_sha":"7149418ed375d2e73256c7bee675c8ce380794ba"} -->
